File: rust/remote_access_tests/src/mock_server.rs
use std::sync::Arc;
use std::time::Duration;

use axum::extract::{Query, State};
use axum::http::{HeaderMap, StatusCode};
use axum::response::sse::{Event, KeepAlive, Sse};
use axum::{Json, Router};
use futures_util::stream;
use serde::{Deserialize, Serialize};
use tokio::net::TcpListener;

use crate::livekit_token;

pub const TEST_DEVICE_TOKEN: &str = "fox_dt_testtoken";
pub const TEST_DEVICE_ID: &str = "dev_testdevice";
const TEST_DEVICE_NAME: &str = "test-device";
const TEST_PROJECT_ID: &str = "prj_testproj";

/// Heartbeat cadence advertised by the mock to a connecting gateway. Short enough that lost
/// heartbeats are detected quickly during tests, but long enough that the SSE read-timeout
/// (a multiple of this) does not race normal test setup.
const MOCK_HEARTBEAT_INTERVAL_MS: u64 = 5_000;
/// Idle timeout advertised by the mock. Picked large so existing tests, which leave viewers
/// connected for a few seconds at most, do not trigger a session-end transition.
const MOCK_DEVICE_WAIT_FOR_VIEWER_MS: u64 = 300_000;

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct DeviceResponse {
    id: String,
    name: String,
    project_id: String,
    retain_recordings_seconds: Option<u64>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct WatchHello {
    watch_lease_id: String,
    device_wait_for_viewer_ms: u64,
    heartbeat_interval_ms: u64,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct WatchWake {
    remote_access_session_id: String,
    url: String,
    token: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct HeartbeatBody {
    #[serde(default)]
    #[allow(dead_code)]
    watch_lease_id: String,
}

#[derive(Deserialize)]
#[serde(rename_all = "camelCase")]
struct WatchQueryParams {
    #[serde(default)]
    #[allow(dead_code)]
    protocol_version: Option<String>,
    #[serde(default)]
    #[allow(dead_code)]
    remote_access_session_id: Option<String>,
    #[serde(default)]
    #[allow(dead_code)]
    previous_watch_lease_id: Option<String>,
}

struct MockState {
    room_name: String,
}

pub struct MockServerHandle {
    url: String,
    join_handle: tokio::task::JoinHandle<()>,
}

impl MockServerHandle {
    pub fn url(&self) -> &str {
        &self.url
    }
}

impl Drop for MockServerHandle {
    fn drop(&mut self) {
        self.join_handle.abort();
    }
}

/// Starts a mock Foxglove API server that emits a watch stream wake carrying a LiveKit
/// access token for the local dev server, and accepts heartbeats.
pub async fn start_mock_server(room_name: &str) -> MockServerHandle {
    let state = Arc::new(MockState {
        room_name: room_name.to_string(),
    });

    let app = Router::new()
        .route(
            "/internal/platform/v1/device-info",
            axum::routing::get(device_info_handler),
        )
        .route(
            "/internal/platform/v1/remote-sessions/watch",
            axum::routing::get(watch_handler),
        )
        .route(
            "/internal/platform/v1/remote-sessions/watch/heartbeat",
            axum::routing::post(heartbeat_handler),
        )
        .with_state(state);

    let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
    let addr = listener.local_addr().unwrap();
    let join_handle = tokio::spawn(async move { axum::serve(listener, app).await.unwrap() });

    MockServerHandle {
        url: format!("http://{addr}"),
        join_handle,
    }
}

fn validate_device_token(headers: &HeaderMap) -> Result<(), StatusCode> {
    let auth = headers
        .get("Authorization")
        .and_then(|v| v.to_str().ok())
        .ok_or(StatusCode::UNAUTHORIZED)?;
    if auth != format!("DeviceToken {TEST_DEVICE_TOKEN}") {
        return Err(StatusCode::UNAUTHORIZED);
    }
    Ok(())
}

async fn device_info_handler(headers: HeaderMap) -> Result<Json<DeviceResponse>, StatusCode> {
    validate_device_token(&headers)?;
    Ok(Json(DeviceResponse {
        id: TEST_DEVICE_ID.into(),
        name: TEST_DEVICE_NAME.into(),
        project_id: TEST_PROJECT_ID.into(),
        retain_recordings_seconds: Some(3600),
    }))
}

/// Emits a `hello` event followed immediately by a `wake` event carrying a freshly-minted
/// LiveKit token for the configured room. After the wake the gateway closes its end of the
/// stream, so we don't need to keep it open.
async fn watch_handler(
    Query(_query): Query<WatchQueryParams>,
    State(state): State<Arc<MockState>>,
    headers: HeaderMap,
) -> Result<Sse<impl futures_util::Stream<Item = Result<Event, axum::Error>>>, StatusCode> {
    validate_device_token(&headers)?;
    let token = livekit_token::generate_token(&state.room_name, TEST_DEVICE_ID)
        .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?;

    let lease_id = format!(
        "rwl_{}",
        std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .unwrap()
            .as_nanos()
    );
    let hello = WatchHello {
        watch_lease_id: lease_id,
        device_wait_for_viewer_ms: MOCK_DEVICE_WAIT_FOR_VIEWER_MS,
        heartbeat_interval_ms: MOCK_HEARTBEAT_INTERVAL_MS,
    };
    let wake = WatchWake {
        remote_access_session_id: "ras_0000mockSession".into(),
        url: livekit_token::livekit_url(),
        token,
    };

    let events = vec![
        Event::default()
            .event("hello")
            .json_data(&hello)
            .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?,
        Event::default()
            .event("wake")
            .json_data(&wake)
            .map_err(|_| StatusCode::INTERNAL_SERVER_ERROR)?,
    ];
    let body_stream = stream::iter(events.into_iter().map(Ok));
    Ok(Sse::new(body_stream).keep_alive(KeepAlive::new().interval(Duration::from_secs(15))))
}

async fn heartbeat_handler(
    headers: HeaderMap,
    Json(_body): Json<HeartbeatBody>,
) -> Result<StatusCode, StatusCode> {
    validate_device_token(&headers)?;
    Ok(StatusCode::OK)
}

File: cpp/foxglove/tests/integration/mock_server.cpp
#include "mock_server.hpp"

#include <nlohmann/json.hpp>

#include <atomic>
#include <chrono>
#include <httplib.h>
#include <sstream>
#include <string>
#include <thread>

#include "livekit_token.hpp"

namespace foxglove_integration {

namespace {

constexpr const char* TEST_DEVICE_NAME = "test-device";
constexpr const char* TEST_PROJECT_ID = "prj_testproj";

bool validate_device_token(const httplib::Request& req) {
  auto it = req.headers.find("Authorization");
  if (it == req.headers.end()) {
    return false;
  }
  return it->second == std::string("DeviceToken ") + TEST_DEVICE_TOKEN;
}

}  // namespace

MockServerHandle::MockServerHandle(const std::string& room_name)
    : server_(std::make_unique<httplib::Server>())
    , stop_flag_(std::make_shared<std::atomic<bool>>(false)) {
  std::string room = room_name;
  auto stop_flag = stop_flag_;

  server_->Get(
    "/internal/platform/v1/device-info",
    [](const httplib::Request& req, httplib::Response& res) {
      if (!validate_device_token(req)) {
        res.status = 401;
        return;
      }
      nlohmann::json body = {
        {"id", TEST_DEVICE_ID},
        {"name", TEST_DEVICE_NAME},
        {"projectId", TEST_PROJECT_ID},
        {"retainRecordingsSeconds", 3600},
      };
      res.set_content(body.dump(), "application/json");
    }
  );

  server_->Get(
    "/internal/platform/v1/remote-sessions/watch",
    [room, stop_flag](const httplib::Request& req, httplib::Response& res) {
      if (!validate_device_token(req)) {
        res.status = 401;
        return;
      }

      auto token = generate_token(room, TEST_DEVICE_ID);
      auto now_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(
                      std::chrono::system_clock::now().time_since_epoch()
      )
                      .count();
      std::ostringstream lease_oss;
      lease_oss << "rwl_" << std::hex << now_ns;
      std::string lease_id = lease_oss.str();

      nlohmann::json hello = {
        {"watchLeaseId", lease_id},
        {"deviceWaitForViewerMs", 300000},
        {"heartbeatIntervalMs", 5000},
      };
      nlohmann::json wake = {
        {"remoteAccessSessionId", "ras_0000mockSession"},
        {"url", livekit_url()},
        {"token", token},
      };

      std::string hello_event = "event: hello\ndata: " + hello.dump() + "\n\n";
      std::string wake_event = "event: wake\ndata: " + wake.dump() + "\n\n";

      res.set_chunked_content_provider(
        "text/event-stream",
        [stop_flag, hello_event, wake_event](size_t offset, httplib::DataSink& sink) -> bool {
          if (offset == 0) {
            std::string events = hello_event + wake_event;
            if (!sink.write(events.data(), events.size())) {
              return false;
            }
            return true;
          }
          // Subsequent invocations: emit a keepalive comment every few seconds while the
          // server is alive. The gateway interprets any byte as proof of life.
          for (int i = 0; i < 50; ++i) {
            if (stop_flag->load()) {
              sink.done();
              return false;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
          }
          static const std::string keepalive = ": keepalive\n\n";
          if (!sink.write(keepalive.data(), keepalive.size())) {
            return false;
          }
          return true;
        }
      );
    }
  );

  server_->Post(
    "/internal/platform/v1/remote-sessions/watch/heartbeat",
    [](const httplib::Request& req, httplib::Response& res) {
      if (!validate_device_token(req)) {
        res.status = 401;
        return;
      }
      res.status = 200;
    }
  );

  int port = server_->bind_to_any_port("127.0.0.1");
  url_ = "http://127.0.0.1:" + std::to_string(port);

  thread_ = std::thread([this]() {
    server_->listen_after_bind();
  });
}

MockServerHandle::~MockServerHandle() {
  if (stop_flag_) {
    stop_flag_->store(true);
  }
  if (server_) {
    server_->stop();
  }
  if (thread_.joinable()) {
    thread_.join();
  }
}

MockServerHandle::MockServerHandle(MockServerHandle&& other) noexcept
    : server_(std::move(other.server_))
    , thread_(std::move(other.thread_))
    , url_(std::move(other.url_))
    , stop_flag_(std::move(other.stop_flag_)) {}

MockServerHandle& MockServerHandle::operator=(MockServerHandle&& other) noexcept {
  if (this != &other) {
    if (stop_flag_) {
      stop_flag_->store(true);
    }
    if (server_) {
      server_->stop();
    }
    if (thread_.joinable()) {
      thread_.join();
    }
    server_ = std::move(other.server_);
    thread_ = std::move(other.thread_);
    url_ = std::move(other.url_);
    stop_flag_ = std::move(other.stop_flag_);
  }
  return *this;
}

MockServerHandle start_mock_server(const std::string& room_name) {
  return MockServerHandle(room_name);
}

}  // namespace foxglove_integration
